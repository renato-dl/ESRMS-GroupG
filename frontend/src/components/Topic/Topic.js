import React from 'react';
import { api } from '../../services/api';
import './Topic.scss';
import {
  Table,
  Button,
  Icon,
  Container
} from 'semantic-ui-react'
import moment from 'moment';
import TopicDetails from './TopicDetail/TopicDetails';
import TopicDelete from './TopicDetail/TopicDelete';
import { NoData } from '../NoData/NoData';

export class Topic extends React.Component{
    constructor(props) {
      super(props);
      this.state = {
        subject: null,
        topics: [],
        editingTopic: null,
        deleteTopicOpen: false, 
        selectedTopic: null,
        isTopicDetailsOpen: false
      }
    }
    
    async componentDidMount() {
      await this.fetchTopics();
      await this.fetchSubject();
    }

    fetchTopics =  async () => {
      const {params} = this.props.match;
      
      //ATTENTION CLASS ID IS HARDCODED
      const response = await api.teacher.getTeacherTopics( 1, params.subjectID);
      if (response) {
        this.setState({ topics: response.data, editingTopic: null })
      } 
    }; 

    fetchSubject = async () => {
      const {params} = this.props.match;

      const response = await api.teacher.getTeacherSubjects(params.teacherID);
      if (response) {
        response.data.forEach((subject) => {
          if (subject.subjectId === params.subjectID) {
            this.setState({subject});
          }
        });
      }
    };

    addTopic = () => {
      this.setState({isTopicDetailsOpen: true});
    };

    onTopicDetailClose = () => {
      this.setState({editingTopic: null, isTopicDetailsOpen: false});
    };

    editTopic = (topic) => {
      this.setState({editingTopic: topic, isTopicDetailsOpen: true});
    };

    // open modal for deleting topic
    deleteTopic = (topic) =>{
      this.setState({deleteTopicOpen: true, selectedTopic: topic});
    }
    onDeleteTopicClose = () =>{
      this.setState({deleteTopicOpen: false, selectedTopic: null});
    }

    // async deleteTopic (topic) {
    //  const response=await api.teacher.deleteTopic(topic)
    //  if (response) {
    //   this.fetchTopics();
    //  } 
    
    // };


    render(){
      if (this.state.topics.length){

      return (
        <Container className="Topic-container contentContainer">
          <h3 className="contentHeader">
            <Icon name='braille'/>
            {this.state.subject ? this.state.subject.subject : ''} topics
          </h3>
          <Button className="ui vk button" onClick={this.addTopic}>
            <Icon name="plus" />
            Add topic
          </Button>
          <Table celled columns={4}>
            <Table.Header>
              <Table.Row>
                <Table.HeaderCell textAlign="left">#</Table.HeaderCell>
                <Table.HeaderCell textAlign="left">Title</Table.HeaderCell>
                <Table.HeaderCell textAlign="left">Description</Table.HeaderCell>
                <Table.HeaderCell textAlign="left">Date</Table.HeaderCell>
                <Table.HeaderCell textAlign="left">Actions</Table.HeaderCell>
              </Table.Row>
            </Table.Header>
            <Table.Body>
              {this.state.topics.map((topic, index) =>
                <Table.Row key={index}>
                  <Table.Cell textAlign="left" width={1}>{ index + 1 }</Table.Cell>
                  <Table.Cell textAlign="left">{ topic.Title }</Table.Cell>
                  <Table.Cell textAlign="left">{ topic.TopicDescription }</Table.Cell>
                  <Table.Cell textAlign="left" width={2}>{ moment(topic.TopicDate).format('LL') }</Table.Cell>
                  <Table.Cell textAlign="left" className="edit-cell" width={1}>
                    <Icon name="edit" onClick={() =>this.editTopic(topic)}/> Edit
                    <br/>
                    <Icon name="delete" onClick={()=>this.deleteTopic(topic)}/> Delete
                  </Table.Cell>
                </Table.Row>
              )}
            </Table.Body>
          </Table>
          {this.state.isTopicDetailsOpen &&
            <TopicDetails
              topic={this.state.editingTopic}
              onClose={this.onTopicDetailClose}
              onSave={() => {
                this.fetchTopics();
                this.onTopicDetailClose();
              }}
            />
          }
          {this.state.deleteTopicOpen &&
            <TopicDelete
              topic={this.state.selectedTopic}
              onClose={this.onDeleteTopicClose}
              onSave={() =>{
                this.fetchTopics();
                this.onDeleteTopicClose();
              }}
            />
          }
        </Container>
      );

      }

      return(
        <Container className="Topic-container contentContainer">
          <h3 className="contentHeader">
            <Icon name='braille'/>
            {this.state.subject ? this.state.subject.subject : ''} topics
          </h3>
          <Button className="ui vk button" onClick={this.addTopic}>
            <Icon name="plus" />
            Add topic
          </Button>


          <NoData/>
          {this.state.isTopicDetailsOpen &&
            <TopicDetails
              topic={this.state.editingTopic}
              onClose={this.onTopicDetailClose}
              onSave={() => {
                this.fetchTopics();
                this.onTopicDetailClose();
              }}
            />
          }
        </Container>
      );
    }
}
