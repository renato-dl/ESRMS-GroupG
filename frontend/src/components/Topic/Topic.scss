.Topic-container {
 .edit-cell {
   &:hover {
     cursor: pointer;
   }
 }
}