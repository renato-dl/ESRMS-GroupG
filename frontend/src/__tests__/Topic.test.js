import React from 'react';
import {Topic} from '../components/Topic/Topic';
import {getRouterPropsForTest} from '../utils';
import {shallow} from 'enzyme';

describe('Testing Topic component', () => {
  
  test('Test if component is rendered', () => {
    shallow(
      <Topic {...getRouterPropsForTest()} />
    );
  });

});
