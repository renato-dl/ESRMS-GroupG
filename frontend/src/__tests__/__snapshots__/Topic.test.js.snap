// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Topic component Testing snapshot 1`] = `
<div
  className="Topic-container contentContainer"
>
  <h3
    className="contentHeader"
  >
    <i
      aria-hidden="true"
      className="braille small icon"
      onClick={[Function]}
    />
    
     topics:
  </h3>
  <button
    className="ui button ui vk button"
    onClick={[Function]}
  >
    <i
      aria-hidden="true"
      className="plus icon"
      onClick={[Function]}
    />
    Add topic
  </button>
  <table
    className="ui celled table"
  >
    <thead
      className=""
    >
      <tr
        className=""
      >
        <th
          className="left aligned"
        >
          #
        </th>
        <th
          className="left aligned"
        >
          Title
        </th>
        <th
          className="left aligned"
        >
          Description
        </th>
        <th
          className="left aligned"
        >
          Date
        </th>
        <th
          className="left aligned"
        >
          Actions
        </th>
      </tr>
    </thead>
    <tbody
      className=""
    />
  </table>
</div>
`;
